torch/distributed/_sharded_tensor/api.py

```diff
@@ -287,13 +287,11 @@ def gather(
 
         world_size = dist.get_world_size(self._process_group)
 
-        gathered_shards = [None] * world_size
-        # will revise this part with CPU support and use dist.gather()
-        # once NCCL support for gather() is ready
-        # https://github.com/pytorch/pytorch/issues/66187
-        dist.all_gather_object(
+        gathered_shards: List[Optional[List[Shard]]] = [None] * world_size if rank == dst else []
+        dist.gather_object(
```

Comment:
(no need to update current PR, can be done in followup ones)

Not sure how performance critical this is, but as we discussed in today's meeting, this indeed looks more expensive than necessary, as there will be additional H2D + D2H copies. I'd assume handling tensor and non-tensor parts separately would be faster for large ShardedTensors. 

https://github.com/pytorch/pytorch/blob/03f1f0cfe45d959a0257aca729239d2c62c4be20/torch/distributed/distributed_c10d.py#L1551-L1565

BTW, is the non-tensor meta info static? If so, we can cache those?

Reply:
Yeah thanks for suggestion, will try to use two separate gather to improve the perf. The non-tensor meta info might not be static I think (i.e. if we do resharding on a ShardedTensor, the Shard.metadata might get changed, to different ranks, or the shard_offset/size changes).

Reply:
I am thinking of a new way to possibly only requires one `gather` call, this requires us make `Shard` a subclass of `torch.Tensor`(metadata is a field in python), so that we can do gather alone. But I am not sure if our c10d collectives support custom tensor objects? (Maybe not as we eventually lowering the collective to C++ and we might only have the at::Tensor do the real communication, not the metadata)

Reply:
> But I am not sure if our c10d collectives support custom tensor objects?

It should. At least it worked for SparseTensor. But I am not sure if that's sufficient for ShardedTensor

https://github.com/pytorch/pytorch/blob/7beb030e11cbaa0c0ebe8d15dce5efce74fc8ae3/torch/csrc/distributed/c10d/ProcessGroupGloo.cpp#L1050-L1134

Reply:
I gave a try on using two separate gather for metadata and tensors to avoid the pickling copies with `gather_object` today. I think it's more tricky than I thought, mainly because:
1. We can use `gather_object` for metadatas, but we couldn't simply use `gather` for the local shard tensors, mainly because `local_shards()` is a list of tensor, but input of `gather` is a single tensor not a list of tensor, so `gather` won't work here
2. We can try using `torch.cat` locally on each rank to form a single tensor before the `gather` collective, but we need to split them afterwards as they might not be adjacent to each other (local shards on this rank might contain two tensors that's far away in the global position logically). So we might need to insert additional insertion point in the first `gather_object` to split the combined tensors from the second `gather` call. This is pretty tricky from my understanding. 

Any suggestions are appreciated :)

For this PR itself, I think we can land it as is, as the code before this PR used `all_gather_object` anyways. I will think more about how to solve this pickling issue and make a follow up PR to improve the perf.

Reply:
@wanchaol Can we create a gh issue to track this follow up improvement?